### PR TITLE
[DO NOT MERGE] multiplayer version 4: part 4: room descriptions

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
**DO NOT MERGE OR PUT THIS ON CANARY. THIS REQUIRES A VERSION BUMP.**
When this is reviewed, I (or the maintainers) can push it to #4377.
To allow for easier compare, visit https://github.com/zhaowenlan1779/citra/compare/multiplayer-v4/main...zhaowenlan1779:multiplayer-v4/feature/room-description

---

Remake of #3721. Please look at the original PR for description.

Changes to the code
---
* Cleaned up the code and removed unnecessary changes
* Saved room description in UISettings like the other fields
* Minor UI changes. See screenshot.

Screenshot
---
image

View the original PR for more screenshots.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4393)

### CODE REVIEW

**Comment by zhaowenlan1779** [comment]
@FearlessTobi I'm not sure what I can do here either. If anyone had a suggestion, please leave a comment.

**Comment by adityaruplaha** [comment]
As an extra newline & make the heading bold, maybe?

>**DESCRIPTION:**
This is a sample description.

>Adityarup_Laha is not playing a game

File: src/citra_qt/multiplayer/lobby_p.h [hunk 1]
```diff
@@ -55,6 +55,31 @@ class LobbyItemName : public LobbyItem {
...
+    }
+
+    QVariant data(int role) const override {
+        if (role != Qt::DisplayRole) {
+            return LobbyItem::data(role);
+        }
+        auto description = data(DescriptionRole).toString();
+        description.prepend("Description: ");
+        return description;
+    }
```
Comment (adityaruplaha) [inline src/citra_qt/multiplayer/lobby_p.h]: You should return "" is description is absent.

File: src/web_service/announce_room_json.cpp [hunk 2]
```diff
@@ -41,6 +42,11 @@ void to_json(nlohmann::json& json, const Room& room) {
 void from_json(const nlohmann::json& json, Room& room) {
     room.ip = json.at("address").get<std::string>();
     room.name = json.at("name").get<std::string>();
+    try {
+        room.description = json.at("description").get<std::string>();
+    } catch (const nlohmann::detail::out_of_range& e) {
+        LOG_DEBUG(Network, "Room \'{}\' doesn't contain a description", room.name);
+    }
```
Comment (adityaruplaha) [inline src/web_service/announce_room_json.cpp]: Note:
The `try {} catch () {}` has to be removed after this is merged.

File: src/network/room.h [hunk 3]
```diff
@@ -57,7 +58,8 @@ enum RoomMessageTypes : u8 {
     IdMacCollision,
     IdVersionMismatch,
     IdWrongPassword,
-    IdCloseRoom
+    IdCloseRoom,
+    IdRoomIsFull,
```
Comment (adityaruplaha) [inline src/network/room.h]: Nit:
Extra `,` at end.

Comment (zhaowenlan1779) [reply src/network/room.h]: I believe it's actually preferred.

File: src/citra_qt/multiplayer/lobby_p.h (hunk 1 above)
Comment (zhaowenlan1779) [reply src/citra_qt/multiplayer/lobby_p.h]: No, if it's empty, `LobbyItemDescription` won't even be displayed.

**Comment by zhaowenlan1779** [comment]
> As an extra newline & make the heading bold & caps, maybe?

To be rather honest I think that will make it look even worse....

**Comment by zhaowenlan1779** [comment]
Pushed to #4377.

**Pull request closed without merging**